File: lua/problem_35.lua
local function sortRGB(arr)
    local lo, mid, hi = 1, 1, #arr
    while mid <= hi do
        if arr[mid] == 'R' then
            arr[lo], arr[mid] = arr[mid], arr[lo]
            lo = lo + 1; mid = mid + 1
        elseif arr[mid] == 'G' then
            mid = mid + 1
        else
            arr[mid], arr[hi] = arr[hi], arr[mid]
            hi = hi - 1
        end
    end
end

local function tableEq(a, b)
    if #a ~= #b then return false end
    for i = 1, #a do if a[i] ~= b[i] then return false end end
    return true
end

local function problem35()
    print("Problem 35:")
    local arr = {'G','B','R','R','B','R','G'}
    sortRGB(arr)
    assert(tableEq(arr, {'R','R','R','G','G','B','B'}), "assertion failed")
    local arr2 = {'B','G','R'}
    sortRGB(arr2)
    assert(tableEq(arr2, {'R','G','B'}), "assertion failed")
    print("All Tests Passed")
end

return { problem35 = problem35 }

File: python/problem_35.py
from typing import List


def sort_rgb(arr: List[str]) -> None:
    lo, mid, hi = 0, 0, len(arr) - 1
    while mid <= hi:
        if arr[mid] == 'R':
            arr[lo], arr[mid] = arr[mid], arr[lo]
            lo += 1; mid += 1
        elif arr[mid] == 'G':
            mid += 1
        else:
            arr[mid], arr[hi] = arr[hi], arr[mid]
            hi -= 1


def main():
    print("Problem 35:")
    arr = ['G', 'B', 'R', 'R', 'B', 'R', 'G']
    sort_rgb(arr)
    assert arr == ['R', 'R', 'R', 'G', 'G', 'B', 'B']
    arr2 = ['B', 'G', 'R']
    sort_rgb(arr2)
    assert arr2 == ['R', 'G', 'B']
    arr3 = ['R']
    sort_rgb(arr3)
    assert arr3 == ['R']
    print("All Tests Passed")
